collaborative-plonk/src/round3.rs

```diff
@@ -0,0 +1,602 @@
+            e3b_ = driver.add_with_public(&challenges.gamma, &e3b_);
+            e3b.push(e3b_);
+
+            let mut e3c_ = c;
+            e3c_ = driver.add_with_public(&(s3 * challenges.beta), &e3c_);
+            e3c_ = driver.add_with_public(&challenges.gamma, &e3c_);
+            e3c.push(e3c_);
+
+            let e3d_ = zw;
+            e3d.push(e3d_);
+            w *= pow_plus2_root_of_unity;
+        }
+
+        let e2a_vec = e2a.into();
+        let e2b_vec = e2b.into();
+        let e2c_vec = e2c.into();
+        let e2d_vec = e2d.into();
+        let zp_vec = zp.into();
+
+        let [e2, e2z_0, e2z_1, e2z_2, e2z_3] = mul4vec!(
+            driver, &e2a_vec, &e2b_vec, &e2c_vec, &e2d_vec, &ap_vec, &bp_vec, &cp_vec, &zp_vec,
+            &domain1
+        );
+
+        let e3a_vec = e3a.into();
+        let e3b_vec = e3b.into();
+        let e3c_vec = e3c.into();
+        let e3d_vec = e3d.into();
+        let zwp_vec = zwp.into();
+
+        let [e3, e3z_0, e3z_1, e3z_2, e3z_3] = mul4vec!(
+            driver, &e3a_vec, &e3b_vec, &e3c_vec, &e3d_vec, &ap_vec, &bp_vec, &cp_vec, &zwp_vec,
+            &domain1
+        );
+
+        let mut t_vec = Vec::with_capacity(zkey.domain_size * 4);
+        let mut tz_vec = Vec::with_capacity(zkey.domain_size * 4);
+        for i in 0..zkey.domain_size * 4 {
+            let mut e2 = T::index_sharevec(&e2, i);
+            let mut e2z = mul4vec_post!(driver, e2z_0, e2z_1, e2z_2, e2z_3, i, z1, z2, z3);
+            let mut e3 = T::index_sharevec(&e3, i);
+            let mut e3z = mul4vec_post!(driver, e3z_0, e3z_1, e3z_2, e3z_3, i, z1, z2, z3);
+
+            let z = T::index_sharevec(&polys.z.eval, i);
+            let zp = T::index_sharevec(&zp_vec, i);
+
+            e2 = driver.mul_with_public(&challenges.alpha, &e2);
+            e2z = driver.mul_with_public(&challenges.alpha, &e2z);
+
+            e3 = driver.mul_with_public(&challenges.alpha, &e3);
+            e3z = driver.mul_with_public(&challenges.alpha, &e3z);
+
+            let mut e4 = driver.add_with_public(&-P::ScalarField::one(), &z);
+            e4 = driver.mul_with_public(&zkey.lagrange[0].evaluations[i], &e4);
+            e4 = driver.mul_with_public(&challenges.alpha2, &e4);
+
+            let mut e4z = driver.mul_with_public(&zkey.lagrange[0].evaluations[i], &zp);
+            e4z = driver.mul_with_public(&challenges.alpha2, &e4z);
+
+            let mut t = driver.add(&e1[i], &e2);
+            t = driver.sub(&t, &e3);
+            t = driver.add(&t, &e4);
+
+            let mut tz = driver.add(&e1z[i], &e2z);
+            tz = driver.sub(&tz, &e3z);
+            tz = driver.add(&tz, &e4z);
+
+            t_vec.push(t);
+            tz_vec.push(tz);
+        }
+        let mut coefficients_t = driver.ifft(&t_vec.into(), &domains.extended_domain);
+        driver.neg_vec_in_place_limit(&mut coefficients_t, zkey.domain_size);
+
+        for i in zkey.domain_size..zkey.domain_size * 4 {
```

Comment:
same for this loop

Reply:
commented and enforced with for_each loop